kubeflow/manifests
# oauth2-proxy with istio mesh config and m2m bearer tokens

**Which issue is resolved by this Pull Request:**
Resolves #2516, #2517

**Description of your changes:**
Changes are based on discussion and description available here:
* https://github.com/kubeflow/manifests/pull/2409#issuecomment-1694495858
* https://github.com/kubeflow/manifests/issues/2516
* https://github.com/kubeflow/manifests/issues/2517

**TL;DR:**
* Drop EnvoyFilter in oauth2-proxy kustomization
* Use Istio `envoyExtAuthzHttp` instead of EnvoyFilter
* Enable m2m bearer tokens
* Use kustomize components for configuration tailoring
* Enable scenario with in-cluster Kubernetes OIDC served behind self-signed certificates
* Explain new auth routine
* Use oauth2-proxy in logout routine

**Testing setup:**
In `common/oidc-client/oauth2-proxy/README.md` are instructions on how to enable usage of `oauth2-proxy`. This covers the example scenario with Kubeflow deployed on kind, which enables configuration for in-cluster OIDC behind self-signed certificates.

To verify if oauth2-proxy is working for standard user login and m2m tokens:
* try to login to Kubeflow with example user/password. After that, you should be able to see:
    * logged in user is `user@example.com`
    * there is a cookie named `oauth2_proxy_kubeflow` in your browser
* create a Pod with `curl` image and try calling `istio-ingressgateway` with Authorization Bearer Token from Service Account:
    ```bash
    [user@example ~]$ kubectl run -ti --rm --image curlimages/curl curl --command -- sh
    ~ $ curl istio-ingressgateway.istio-system/api/workgroup/exists -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)"
    {"hasAuth":true,"user":"system:serviceaccount:default:default","hasWorkgroup":false,"registrationFlowAllowed":false}
    ```

**Comments:**
@axel7083 , I'd appreciate your feedback and it would be great if you could test this setup

@juliusvonkohout , I know the file structure looks a little bit different but I thought it makes a little bit more sense. That said, I'm open for discussion and we can change the file structure.

From implementation perspective everything is in place. The only missing parts are adding some more README, probably doing some clean up and addressing feedback.

I'd appreciate review.

---

## Discovered missing points

- [x] enable oauth2-proxy as default
- [x] enable m2m by default
- [x] enable self-signed certs by default
- [x] write e2e tests for m2m - https://github.com/kubeflow/manifests/pull/2417
  - [x] list pipeline runs
  - [x] list experiments in a namespace
  - [x] starting a pipeline - good to have if time allows, might require considerate amount of work
  - [x] test kserve (create new e2e github workflow based on the current kserve test)
    - use only the scenario with standard istio installation (not cni)
- [x] HA for oauth2-proxy with multiple replicas
- [x] describe oauth2-proxy upgrade instructions
